### PR TITLE
property types support

### PR DESCRIPTION
**Pull request by basdewitte**
Hi Martin, thanks for sharing your code four years ago. I fixed support for non string properties such as Long/long and Integer/int. The changes are very straightforward. See the junit test for details. If you agree this is an improvement I'd welcome you to make it available on maven central so I can use the public version in my own project.

Furthermore I would like to make some more changes: 

1) Do not ignore defaultValues from annotation if properties file is missing. Currently the defaults are ignored unless the file exists. IMHO this is a bug. 2) Environment variable values are also ignored on missing resource which I would like to fix. 

3) Add a setting to optionally throw exception on missing resource. Currently it writes a warning to stdout.  E.g.:
```
PropertyValueInjector.mandatoryInject(myConfig);
or 
PropertyValueInjector.inject(myConfig, true);
```

4) Use pojo property name for key if key omitted in annotation.
```
@PropertyValue()
protected String username;

#properties file
username=carlspring
```
5) Add support for loading the properties from a HOME_VAR defined location. Currently the properties files are only read from a classpath location. I'd like to add support for something like:
``
PropertyValueInjector.inject(myConfig, "CARL_HOME", true);
//
String confDir = System.getProperty(this.confDirVar);
properties.load(Files.newInputStream(Paths.get(confDir, resource)));
``
6) Perhaps add integration with commons configuration? I could make changes to allow for overriding the loadResource() method so one can create a subclass which uses FileLocationStrategy?

Please let me know if you have any interest in the above.

Thanks
Bas

### CODE REVIEW

**Comment by carlspring** [comment]
Hi @basdewitte!

Happy New Year!

I'm happy that someone other than myself has found this useful as well!

I will try and review it tonight, or tomorrow and let you know my thoughts. In the meantime, would you mind reformatting your changes according to the code styla that I've used so far?

Kind regards,

Martin

**Comment by carlspring** [comment]
Hi,

You suggestions all seem quite good! Feel free to contribute pull requests for them! I can make a release when you need me to.

Looking forward to your contributions!

Martin

**Comment by carlspring** [comment]
Let me know when you're ready, so I can review, merge and create a new release for you.

**Comment by basdewitte** [comment]
Go ahead! All straightforward changes are done. See changelog.

Less straightforward  changes I could deal with in a while. Please express any considerations you might have on the introduction of dependecies, overhauling your code, requiring java 1.7 instead of 1.6 and whatever else you might care about in relation to the below.
     
I could add some logging using log4j. 

Furthermore I could add support to locate the resource files in other places than the classpath. For this I can think of a number of suitable strategies. Involving:
* a change of the PropertyValueInjector class such that the method to load the property file can be overloaded, looking at thread safety in the meantime
* Using commons.text.StrSubstitutor to expand resources such as "${MY_HOME}/conf/app.properties" or:
* passing a commons.configuration2 FileLocationStrategy.

Thanks
Bas

**Comment by basdewitte** [comment]
Oh, and do you agree it is a bug, in the example below, the bean property _unassigned_ is injected with the value from the other properties resource? Someone may have for instance  _mysql.properties_ and _oracle.properties_ containing identical keys so this is not academic. 

```

public class PropertyHolderWithMultipleResouces
{

    @PropertyValue(key = "jdbc.username", resource = "META-INF/properties/jdbc.properties")
    protected String username;

    @PropertyValue(key = "jdbc.username")
    protected String unassigned;
    
    public String getUsername() {
    ...
    }

}
```

File: CHANGELOG [hunk 1]
```diff
@@ -0,0 +1,4 @@
+2018/01/11	1.9.0 basdewitte	use bean property name for property key if omitted in annotation
```
Comment (carlspring) [inline CHANGELOG]: The version is incorrect. The latest versions which was released was `1.7`, which is why we're currently at `1.8-SNAPSHOT` in the `master`. :)

Could you please revert the change to the `pom.xml` file relating to this as well?

File: pom.xml [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
     <groupId>org.carlspring</groupId>
     <artifactId>properties-injector</artifactId>
-    <version>1.8-SNAPSHOT</version>
+    <version>1.9-SNAPSHOT</version>
```
Comment (carlspring) [inline pom.xml]: (As described above, we still haven're release `1.8`, so there's no need to bump the version here yet).

File: src/main/java/org/carlspring/ioc/PropertyValueInjector.java [hunk 3]
```diff
@@ -259,45 +277,87 @@ private static void setField(Field field,
...
-            // Handle objects
-            else
+            else if (targetType.equals(Character.class) || targetType.equals(Character.TYPE))
             {
-                field.set(target, value);
+                convertedValue = value.toString().charAt(0);
             }
+
+            field.set(target, convertedValue);
+
```
Comment (carlspring) [inline src/main/java/org/carlspring/ioc/PropertyValueInjector.java]: You can delete this empty line.

File: src/main/java/org/carlspring/ioc/PropertyValueInjector.java [hunk 4]
```diff
@@ -259,45 +277,87 @@ private static void setField(Field field,
...
         }
         catch (IllegalAccessException iae)
+        {
+            throw new IllegalArgumentException("Could not set field " + field, iae);
+
```
Comment (carlspring) [inline src/main/java/org/carlspring/ioc/PropertyValueInjector.java]: ... And this empty line.

File: src/test/java/org/carlspring/ioc/mock/PropertyHolderWithIntAndLongProperties.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    protected long primLong;
+
+    @PropertyValue(key = "prim.double")
+    protected double primDouble;
+
+    @PropertyValue(key = "prim.float")
+    protected float primFloat;
+
+    @PropertyValue(key = "prim.bool")
+    protected boolean primBool;
```
Comment (carlspring) [inline src/test/java/org/carlspring/ioc/mock/PropertyHolderWithIntAndLongProperties.java]: This is just a code style remark really (it had nothing to do with the quality of your work) -- I normally use camel hump variables with full names... So this would be `primitiveBoolean`. This makes sure that there no confusions. Alternatively, in methods, if it's very clear what the name is, I do things like `InputStream is = ...`.

In this case, half of the ones are with complete names, the others are abbreviated.... This is just for consistency. :)

**Comment by carlspring** [comment]
I've gone over the pull once again! It looks great. I have a couple of really minor remarks. It would be great, if you could look into those as well.

**Comment by carlspring** [comment]
Also, I agree with your question/suggestion about keys from multiple properties files.

**Comment by carlspring** [comment]
Thanks a lot for these improvements! :)
Ping me when you'd like to have the release cut and deployed to Maven Central.
